### PR TITLE
io: add a BufferedWriter and supporting methods

### PR DESCRIPTION
**Pull request by einar-hjortdal**
Address #21975

### CODE REVIEW

File: vlib/io/buffered_writer_test.v [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+pub mut:
+	result []u8
+}
+
+// implement io.Writer
+fn (mut aw ArrayWriter) write(buf []u8) !int {
+	len := buf.len
+	mut res := 0
+	for i := 0; i < len; i++ {
+		aw.result = arrays.concat(aw.result, buf[i])
```
Comment (spytheman) [inline vlib/io/buffered_writer_test.v:14]:
```suggestion
		aw.result << buf[i]
```

File: vlib/io/buffered_writer_test.v [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		res++
+	}
+	return res
+}
+
+fn create_data(n int) []u8 {
+	mut res := []u8{}
+	for i := 0; i < n; i++ {
+		res = arrays.concat(res, 88) // x
```
Comment (spytheman) [inline vlib/io/buffered_writer_test.v:23]:
```suggestion
		res << `x`
```

File: vlib/io/buffered_writer_test.v [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	res = arrays.concat(res, 10) // \n
```
Comment (spytheman) [inline vlib/io/buffered_writer_test.v:25]:
```suggestion
	res << `\n`
```

File: vlib/io/buffered_writer.v [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	cap    int = 128 * 1024
+}
+
+pub fn new_buffered_writer(o BufferedWriterConfig) !&BufferedWriter {
+	if o.cap < 1 {
+		return error('`o.cap` must be a positive integer')
+	}
+
+	return &BufferedWriter{
+		buf: []u8{len: o.cap, cap: o.cap}
```
Comment (spytheman) [inline vlib/io/buffered_writer.v:24]:
```suggestion
		buf: []u8{len: o.cap}
```
(setting cap: is implied by setting len:)

File: vlib/io/buffered_writer.v [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		wr:  o.writer
+	}
+}
+
+pub fn (mut b BufferedWriter) reset() {
+	cap := b.buf.len
+	b.buf = []u8{len: cap, cap: cap}
```
Comment (spytheman) [inline vlib/io/buffered_writer.v:31]:
```suggestion
	b.buf.reset()
```

File: vlib/io/buffered_writer.v [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		mut n := 0
+		if b.buffered() == 0 {
+			n = b.wr.write(p)!
+		} else {
+			n = copy(mut b.buf[b.n..], p)
+			b.n += n
+			b.flush()!
+		}
+		nn += n
+		p = p[n..].clone()
```
Comment (spytheman) [inline vlib/io/buffered_writer.v:71]: why are you cloning here?
`p` is already a clone of `src`; imho you can freely just slice it with
`p = unsafe{ p[n..] }`, since you do not modify the data in it all in the loop

Comment (spytheman) [reply vlib/io/buffered_writer.v:71]: In fact the first clone at the start, is also not necessary imho, if `p` is going to just get sliced thinner and thinner later.

Comment (einar-hjortdal) [reply vlib/io/buffered_writer.v:71]: I got the
```
Details: vlib/io/buffered_writer.v:71:8: details: To silence this notice, use either an explicit `a[..].clone()`,
or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.
   69 |         }
   70 |         nn += n
   71 |         p = p[n..]
      |              ~~~~~
   72 |     }
   73 |
```
I went for clone() because I try to avoid unsafe blocks.

Comment (spytheman) [reply vlib/io/buffered_writer.v:71]: imho saving the memory allocations and the copying is a better goal, for a library module. Using `unsafe{}` is not a taboo (and V is not Rust). `unsafe{}` blocks just signal to both the developers and the compiler, that what is inside should be reviewed more carefully.

Comment (spytheman) [reply vlib/io/buffered_writer.v:71]: It can be optimized out in a later PR though, if you are not comfortable with changing it. It is not a big deal.

File: vlib/io/buffered_writer.v (hunk 5 above)
Comment (einar-hjortdal) [reply vlib/io/buffered_writer.v:31]: I think this requires an unsafe block

**Comment by einar-hjortdal** [comment]
oops. I messed up the rebase.

File: vlib/io/buffered_writer.v [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	return b.buf.len - b.n
+}
+
+// write writes `src` in the buffer, flushing it to the underlying writer as needed, and returns the
+// number of bytes written.
+pub fn (mut b BufferedWriter) write(src []u8) !int {
+	mut p := src.clone()
+	len := p.len
+	mut nn := 0
+	for len > b.available() {
```
Comment (spytheman) [inline vlib/io/buffered_writer.v:71]: What will happen, if `len` is bigger than the buffer size?
Would not that loop turn into an infinite one?

Comment (spytheman) [reply vlib/io/buffered_writer.v:71]: I think that the loop condition should use `p.len` instead of `len` 🤔 .

File: vlib/io/buffered_writer.v [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+		mut n := 0
+		if b.buffered() == 0 {
+			n = b.wr.write(p)!
```
Comment (spytheman) [inline vlib/io/buffered_writer.v:74]: That does not seem right ... if there is nothing buffered yet, it will just skip the buffer entirely, and write directly to the writer.
What is the point of the buffering then?

Comment (spytheman) [reply vlib/io/buffered_writer.v:74]: I think that in general, buffering provides a way to decouple the speeds, at which producers and consumers of data work.

A buffered writer, should allow a fast producer of data, to fill a large buffer quickly and/or in bursts, and then drain that buffer, at whatever pace the consumer of the data can handle.

Thus in the write() method, if there is still space available, the data should go into the buffer, without concerning with `b.wr` at all, and *only* when the buffer is filled, it should be drained with a loop doing b.wr.write() calls.

File: vlib/io/buffered_writer.v (hunk 7 above)
Comment (einar-hjortdal) [reply vlib/io/buffered_writer.v:71]: You are correct. This will be fixed and a test will be added.

File: vlib/io/buffered_writer.v (hunk 8 above)
Comment (einar-hjortdal) [reply vlib/io/buffered_writer.v:74]: Ideally, the buffer size should be big enough to handle large writes, writes should be done manually using the flush method.
When the write is bigger than the buffer, I think writing directly would prevent needlessly copying data from source to buffer, and then give it to the underlying writer.

Comment (spytheman) [reply vlib/io/buffered_writer.v:73]: The check is just `if b.buffered() == 0 {`, it does not take into account how large the block is at all. It will trigger for even small writes, negating the useful feature of 
the buffer *actually buffering*.

It is just buggy and has to be fixed.

Comment (spytheman) [reply vlib/io/buffered_writer.v:73]: I think that a new module should focus on having correct and useful behavior first, and only then with optimizing edge cases, when there are suitable tests to guarantee that the optimizations will not break the correctness and usefulness.

Comment (einar-hjortdal) [reply vlib/io/buffered_writer.v:73]: > The check is just `if b.buffered() == 0 {`, it does not take into account how large the block is at all. It will trigger for even small writes, negating the useful feature of the buffer _actually buffering_.
> 
> It is just buggy and has to be fixed.

The check is inside the `for p.len > b.available()` block, the loop checks if the source data is larger than the available space in the buffer.
Am I misunderstanding you?

Comment (spytheman) [reply vlib/io/buffered_writer.v:73]: Please add a test, that does several small writes in a loop, with a total size < the buffer length. The buffered writer should not call its writer. Then call the flushing method, and it should write all the accumulated data from all small writes.

That test is the *basic functionality*, that a buffered writer should support, to justify its existence in the first place.

File: vlib/io/buffered_writer_test.v [hunk 9]
```diff
@@ -1,5 +1,5 @@
-import arrays
 import io
+import crypto.rand
```
Comment (spytheman) [inline vlib/io/buffered_writer_test.v:2]: Why do you use `crypto.rand` here?

There is no need to exhaust the entropy of the system in an unrelated test like that.
Use either `rand` (which uses a pseudo random generator, that will not block, or exhaust anything), or just string literals.

File: vlib/io/buffered_writer_test.v [hunk 10]
```diff
@@ -31,6 +31,34 @@ fn test_create_buffered_writer() {
...
+fn test_basic_write() {
+	mut aw := ArrayWriter{}
+	max := 65536
+	mut bw := io.new_buffered_writer(writer: aw, cap: max)!
+
+	// write less data than buffer capacity, the underlying writer should receive no data.
+	safe_max := 65536 - 255
+	mut total := 0
+	for total < safe_max {
+		a := rand.read(1)!
```
Comment (spytheman) [inline vlib/io/buffered_writer_test.v:43]: `rand.u8()` or a literal is a better choice

File: vlib/io/buffered_writer_test.v (hunk 9 above)
Comment (einar-hjortdal) [reply vlib/io/buffered_writer_test.v:2]: That was the first approach I could think of to solve this problem, I shall rewrite the test without crypto.rand

**Comment by einar-hjortdal** [comment]
Test is failing sometimes, I think I wrote a bad utility function in the test and just discovered it.

**Comment by spytheman** [comment]
The windows CI failures are unrelated and are already fixed on master.

**Review by spytheman — APPROVED** [review]
Excellent work.

I could not think about any other tests.

File: vlib/io/buffered_writer_test.v [hunk 11]
```diff
@@ -0,0 +1,133 @@
+import io
+import rand
+
+struct ArrayWriter {
```
Comment (spytheman) [inline vlib/io/buffered_writer_test.v:4]:
```suggestion
struct ArrayWriter implements Writer {
```

Comment (spytheman) [reply vlib/io/buffered_writer_test.v:4]: although that could wait a bit more 🤔

File: vlib/io/buffered_writer.v [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+// Returns an error if the writer fails to write all buffered data.
+pub fn (mut b BufferedWriter) flush() ! {
+	if b.buffered() == 0 {
+		return
+	}
+
+	n := b.wr.write(b.buf[0..b.n])!
+	if n < b.n {
+		return error('Writer accepted less bytes than expected without returning any explicit error.')
+	}
```
Comment (spytheman) [inline vlib/io/buffered_writer.v:54]: we could perhaps loop here, till everything is flushed

Comment (spytheman) [reply vlib/io/buffered_writer.v:54]: or there could be a separate flush_all method that does it 🤔
